### PR TITLE
Save ensemble as attribute in SampleContainer

### PR DESCRIPTION
**Pull request by lbluque**
## Summary

* Save the ensemble used to sample as an attributed of  `SampleContainer`.
* Using the ensemble, implemented `get_sampled_structures` to retrieve pymatgen structures from traced occupancies. Addresses #296 
* The json and hdf5 format to save `SampleContainer` had to be modified. Reading sample containers is backward compatible, which allows users to read and possibly update their containers to the new format.

## Checklist

- [X] Code is in the [standard Python style](https://www.python.org/dev/peps/pep-0008/).
      Run [pycodestyle](https://pycodestyle.readthedocs.io/en/latest/) and [flake8](http://flake8.pycqa.org/en/latest/)
      on your local machine.
- [X] Docstrings have been added in the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).
      Run [pydocstyle](http://www.pydocstyle.org/en/2.1.1/index.html) on your code.
- [X] Tests have been added for any new functionality or bug fixes.
- [X] All existing tests pass.

### CODE REVIEW

File: smol/moca/sampler/container.py [hunk 1]
```diff
@@ -126,6 +140,29 @@ def sampling_efficiency(self, discard=0, flat=True):
             efficiency = efficiency.mean()
         return efficiency
 
+    def get_sampled_structures(self, indices=-1):
+        """Get sampled structures for MC steps given by indices.
+
+        Args:
+            indices (list of int or int):
+                A single index or list of indices to obtain sampled structures
```
Comment (qchempku2017) [inline smol/moca/sampler/container.py:148]: Please explain in this docstring what the default indices=-1 will do.

Comment (qchempku2017) [reply smol/moca/sampler/container.py:148]: Also why not have argument "flatten" in this function as well?

**Comment by lbluque** [comment]
Hi @qchempku2017, can you try loading one of your hdf5 sample containers to test these updates?

File: smol/moca/sublattice.py [hunk 2]
```diff
@@ -176,6 +176,14 @@ def get_index(sp, species):
             part_sublattices.append(part_sublatt)
         return part_sublattices
 
+    def __eq__(self, other):
+        """Check equality of sublattices."""
+        if other.site_space == self.site_space and np.array_equal(
+            other.sites, self.sites
+        ):
```
Comment (qchempku2017) [inline smol/moca/sublattice.py:183]: I suppose sublattice encoding needs to be checked as well?

**Review by qchempku2017 — CHANGES_REQUESTED** [review]
Please see change requests in the comments.

File: smol/moca/sampler/container.py (hunk 1 above)
Comment (lbluque) [reply smol/moca/sampler/container.py]: edited the code to include a `flat` keyword argument, and explained behavior in docstring.

File: smol/moca/sublattice.py (hunk 2 above)
Comment (lbluque) [reply smol/moca/sublattice.py:188]: Thanks! added checking encoding as well

**Review by qchempku2017 — APPROVED** [review]
Thanks! Okay to merge.

**Pull request merged**